<!doctype html>
<html class="no-js" lang="en" dir="ltr">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>BpForms: toolkit for concretely describing non-canonical DNA, RNA, and proteins</title>
    <link rel="stylesheet" href="css/foundation-icons.css">
    <link rel="stylesheet" href="css/foundation.css">
    <link rel="stylesheet" href="css/app.css">
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway">
  </head>
  <body>
    <div class="grid-container">
      <div class="grid-x grid-padding-x">
        <div class="large-12 cell">
          <h1><i>BpForms</i>: toolkit for concretely describing non-canonical DNA, RNA, and proteins</h1>
        </div>
      </div>

      <div class="grid-x grid-padding-x blue">
        <div class="large-12 cell">
          <p><i>BpForms</i> is a toolkit for unambiguously describing the primary sequence of DNA, RNA, and proteins, including non-canonical monomeric forms, circular topologies, and crosslinks. By concretely describing the primary sequence of biopolymers, <i>BpForms</i> aims to help epigenomics, transcriptomics, proteomics, systems biology, and synthetic biology researchers share and integrate information about DNA modification, post-transcriptional modification, post-translational modification, expanded genetic codes, and synthetic parts. In particular, <i>BpForms</i> aims to help researchers collaboratively develop whole-cell computational models <a href="https://www.wholecell.org"><i class="fi-link"></i></a>.</p>

          <p><i>BpForms</i> includes a notation for describing biopolymer forms and three consensus alphabets of non-canonical monomeric forms of DNA nucleobases, RNA nucleosides, and protein residues. <i>BpForms</i> also includes four software tools for verifying descriptions of biopolymer forms and calculating properties such as their atomic structure, formula, molecular weight, and charge: this website, a JSON REST API, a command line interface, and a Python API. <i>BpForms</i> is available open-source under the MIT license.</p>
        </div>
     </div>

      <div class="grid-x grid-padding-x blue">
        <div class="large-12 cell">
            <h2><i>BpForms</i> verifier/calculator</h2>
        </div>
        <div class="large-6 cell">
            <a name="form"></a>
            <h3>Enter a biopolymer form</h3>
            <form>
            <label>Select an alphabet
                <select id="alphabet">
                </select>
            </label>

            <label>Enter a sequence of monomeric forms and, optionally, a set of crosslinks. See the documentation below for more information about the syntax.
                <textarea
                    id="monomer_seq_in"
                    placeholder='AC{m2A}CC[id: dI | structure: "O=C1NC=NC2=C1N=CN2"]G'></textarea>
            </label>

            <label>Select the topology
                <select id="circular">
                  <option value=1>Circular</option>
                  <option value=0 selected>Linear</option>
                </select>
            </label>

            <label>Optionally, enter a pH at which to calculate the major the protonation state
                <input type="number" id="ph" placeholder="7.4"/>
            </label>

            <label>Optionally, calculate the major tautomer state at the entered pH
                <select id="major_tautomer">
                  <option value=1>Yes</option>
                  <option value=0 selected>No</option>
                </select>
            </label>


            <div id="warnings" style="color: #da663b;">
            </div>

            <div id="errors" style="color: #da3b60;">
            </div>

            <input id="submit" type="button" class="button" value="Calculate"/>
            </form>
        </div>
        <div class="large-6 cell">
            <h3>Calculated properties of the biopolymer form</h3>
            <form class="out">
                <div id="structure-img" style="text-align: center;"></div>
                <label>Length
                    <input type="number" id="length"/>
                </label>
                <label>Structure (SMILES)
                    <textarea id="structure"></textarea>
                </label>
                <label>Chemical formula
                    <input type="text" id="formula"/>
                </label>
                <label>Molecular weight
                    <input type="number" id="mol_wt"/>
                </label>
                <label>Charge
                    <input type="number" id="charge"/>
                </label>
            </form>
        </div>
      </div>

      <div class="grid-x grid-padding-x orange">
        <a name="notation"></a>
        <div class="large-12 cell">
            <h2><i>BpForms</i> notation</h2>
        </div>

        <div class="large-5 cell notation">
          <h3>Overview</h3>
          <p class="before_list">The <i>BpForms</i> notation extends the IUPAC/IUBMB notation commonly used to represent unmodified DNA, RNA, and proteins to represent non-canonical forms of DNA, RNA, and proteins:</p>
          <ul>
            <li><i>BpForms</i> can represent a wider range of monomeric forms, including monomeric forms that are not described in pre-defined alphabets.</li>
            <li><i>BpForms</i> can capture left and right caps such as 3' poly A tails.</li>
            <li><i>BpForms</i> can capture modifications to the backbones of DNA, RNA, and proteins.</li>
            <li><i>BpForms</i> can capture linear and circular topologies of polymers.</li>
            <li><i>BpForms</i> can capture crosslinks (additional bonds between non-adjacent monomeric forms).</li>
          </ul>

          <p class="before_list">The <i>BpForms</i> notation consists of three parts separated by pipes ("|") (e.g., <tt>A{pSer}Y | circular | crosslink: [...]</tt>):</p>
          <ul>
            <li>A sequence of strings that indicate monomeric forms of an alphabet bonded together via backbone motifs (e.g, <tt>A{pSer}Y</tt>. As described below, <i>BpForms</i> includes six alphabets of hundreds of monomeric forms of DNA, RNA, and proteins, and users can define additional alphabets. Each alphabet also includes a backbone. See additional information below.</li>
            <li>An optional attribute for indicating that the polymer is circular (e.g., <tt>circular</tt>). See additional information below.</li>
            <li>An optional set of crosslinks (e.g., <tt>crosslink: [...]</tt>). See additional information below.</li>
          </ul>

          <p class="before_list">The monomeric forms in the sequence of monomeric forms can be described in three ways:</p>
          <ul>
            <li>Individual characters (e.g, <tt>A</tt>) can be used to indicate monomeric forms in defined alphabets that have single character codes.</li>
            <li>Multiple characters delimited by curly brackets (e.g., <tt>{pSer}</tt>) can be used to indicate monomeric forms in defined alphabets that have multi-character codes.</li>
            <li>Monomeric forms which are not defined in the alphabets can be defined "inline" using multiple attributes delimited by square brackets (e.g., <tt>[id: "m2A" | ...]</tt>). See additional information below.</li>
          </ul>

          <p>In addition to defining monomeric forms, each alphabet also defines the backbone motif involved in the bonds between successive monomeric forms. <i>BpForms</i> distinguishes between monomeric forms and backbones to enable consistency between the DNA, RNA, and protein alphabets and DNAmod, MODOMICS, the RNA Modification Database, and RESID.</p>

          <p>The <i>BpForms</i> grammar <a href="https://raw.githubusercontent.com/KarrLab/bpforms/master/bpforms/grammar.lark"><i class="fi-link"></i></a> is defined in Lark syntax <a href="https://lark-parser.readthedocs.io"><i class="fi-link"></i></a>, which is based on EBNF syntax <a href="https://en.wikipedia.org/wiki/Extended_Backus%E2%80%93Naur_form"><i class="fi-link"></i></a>.</p>
        </div>
        <div class="large-7 cell example">
          <h3>Examples</h3>
          <div class="grid-x grid-padding-x">
            <div class="large-4 cell">
              <h4>DNA</h4>
              <p style="text-align: center;"><tt>{dI}ACGC</tt></p>
              <img src="img/example/form-DNA.svg"
                style="margin-bottom: 0.5rem;"/>
              <p class="form-properties">
                Length: 5<br/>
                Formula: C<sub>48</sub>H<sub>55</sub>N<sub>20</sub>O<sub>30</sub>P<sub>5</sub><br/>
                Molecular weight: 1546.9<br/>
                Charge: -6
              </p>
              <p>Deoxyinosine at the first position</p>
            </div>
            <div class="large-4 cell">
              <h4>RNA</h4>
              <p style="text-align: center;"><tt>AC{9A}GC</tt></p>
              <img src="img/example/form-RNA.svg"
                style="margin-bottom: 0.5rem;"/>
              <p class="form-properties">
                Length: 5<br/>
                Formula: C<sub>48</sub>H<sub>55</sub>N<sub>20</sub>O<sub>35</sub>P<sub>5</sub><br/>
                Molecular weight: 1626.9<br/>
                Charge: -6
              </p>
              <p>Inosine at the third position</p>
            </div>
            <div class="large-4 cell">
              <h4>Protein</h4>
              <p style="text-align: center;"><tt>ACU[id: "U"]C</tt></p>
              <img src="img/example/form-Protein.svg"
                style="margin-bottom: 0.5rem;"/>
              <p class="form-properties">
                Length: 4<br/>
                Formula: C<sub>12</sub>H<sub>22</sub>N<sub>4</sub>O<sub>5</sub>S<sub>2</sub>Se<sub>1</sub><br/>
                Molecular weight: 445.4<br/>
                Charge: 0
              </p>
              <p>L-selenocysteine at the third position</p>
            </div>
          </div>
        </div>

        <div class="large-5 cell notation">
          <a name="alphabets"></a>
          <h3>Alphabets</h3>
          <p style="margin-bottom: 0px;"><i>BpForms</i> includes several alphabets. Each alphabet includes a backbone motif and hundreds of monomeric monomers.</p>
          <ul id="alphabet_descriptions">
          </ul>
          <p>The backbone for the DNA alphabets is a deoxyribose monophosphate. The backbone for the RNA alphabets is phosphate. The backbone for the protein alphabets is hydroxide (in COOH).</p>
          <p>As described above, monomeric forms with single characters codes can be indicated by their codes (e.g., <tt>C</tt>) and monomeric forms with multi-character codes can be indicated by their codes delimited by curly brackets (e.g., <tt>{m2A}</tt>).</p>
        </div>
        <div class="large-7 cell example">
          <h3>Examples</h3>
          <div class="grid-x grid-padding-x">
            <div class="large-4 cell">
              <h4>DNA</h4>
              <p style="text-align: center;"><tt>C{m2A}G{m2C}</tt></p>
              <img src="img/example/alphabet-DNA.svg"
                style="margin-bottom: 0.5rem;"/>
              <p class="form-properties">
                Length: 4<br/>
                Formula: C<sub>39</sub>H<sub>50</sub>N<sub>17</sub>O<sub>24</sub>P<sub>4</sub><br/>
                Molecular weight: 1264.8<br/>
                Charge: -5
              </p>
              <p>2-methyladenine at the second position and 2-methylcytosine at the fourth position</p>
            </div>
            <div class="large-4 cell">
              <h4>RNA</h4>
              <p style="text-align: center;"><tt>A{21C}GC</tt></p>
              <img src="img/example/alphabet-RNA.svg"
                style="margin-bottom: 0.5rem;"/>
              <p class="form-properties">
                Length: 4<br/>
                Formula: C<sub>44</sub>H<sub>60</sub>N<sub>18</sub>O<sub>29</sub>P<sub>4</sub><br/>
                Molecular weight: 1429.0<br/>
                Charge: -4
              </p>
              <p>2-lysidine at the second position</p>
            </div>
            <div class="large-4 cell">
              <h4>Protein</h4>
              <p style="text-align: center;"><tt>{AA0037}E{AA0038}</tt></p>
              <img src="img/example/alphabet-Protein.svg"
                style="margin-bottom: 0.5rem;"/>
              <p class="form-properties">
                Length: 3<br/>
                Formula: C<sub>12</sub>H<sub>18</sub>N<sub>3</sub>O<sub>14</sub>P<sub>2</sub><br/>
                Molecular weight: 490.2<br/>
                Charge: -5
              </p>
              <p>O-phospho-L-serine at the first position and O-phospho-L-threonine at the third position</p>
            </div>
          </div>
        </div>

        <div class="large-5 cell notation">
          <h3>"Inline" monomeric forms</h3>
          <p>Monomeric forms which are not defined in the alphabet can be defined "inline" within the sequence of monomeric forms.</p>

          <p>Inline monomeric forms can be defined by enclosing multiple attributes separated by pipes ("|") in square brackets (e.g., <tt>[id: "a-short-name" | name: "a long name " | ...]</tt>).</p>

          <p class="before_list">The inline monomeric forms support seven types of attributes. See below for more information about each attribute.</p>
          <ul>
            <li><tt>id</tt>, <tt>name</tt>, <tt>synonym</tt>: These attributes indicate human readable names for the monomeric form.</li>
            <li><tt>structure</tt>: This attribute indicates the atomic structure of the monomeric form.</li>
            <li><tt>backbone-bond-atom</tt>, <tt>backbone-displaced-atom</tt>: These attributes indicate the atoms in the structure that participate in bonds with its backbone and the atoms in the structure which are displaced by forming bonds between the monomeric form and its backbone.</li>
            <li><tt>left-bond-atom</tt>, <tt>left-displaced-atom</tt>, <tt>right-bond-atom</tt>, <tt>right-displaced-atom</tt>: These attributes indicate the atoms in the structure that participate in bonds with the preceeding and following monomeric forms in the sequence and the atoms that are displaced by the formation of these bonds.</li>
            <li><tt>delta-mass</tt>, <tt>delta-charge</tt>, <tt>position</tt>: These attributes can be used to describe uncertainty in the structure and location of the monomeric form.</li>
            <li><tt>identifier</tt>: This attribute indicates an entry in a database or ontology which is equivalent to the monomeric form.</li>
            <li><tt>comment</tt>: This attribute can be used to store comments about monomeric forms.</li>
          </ul>

          <p>All of these attributes are optional. However, the <tt class="attribute">structure</tt>, <tt class="attribute">backbone-bond-atom</tt>, <tt class="attribute">backbone-displaced-atom</tt>, <tt class="attribute">left-bond-atom</tt>, <tt class="attribute">left-displaced-atom</tt>, <tt class="attribute">right-bond-atom</tt>, and <tt class="attribute">right-displaced-atom</tt> attributes are required to calculate the atomic structure, chemical formula, molecular weight, and charge of the polymer.</p>
        </div>
        <div class="large-7 cell example">
          <h3>Examples</h3>
          <pre>[id: "dI"
    | name: "hypoxanthine"
    | structure: "O=C1NC=NC2=C1N=CN2"
    ]</pre><br/>
          <pre>[id: "AA0305"
    | name: "N5-methyl-L-arginine"
    | structure: "O=C[C@H](CCCN(C(=[NH2])N)C)[NH3+]"
    | backbone-bond-atom: C2
    | backbone-displaced-atom: H2
    | right-bond-atom: C2
    | left-bond-atom: N15-1
    | left-displaced-atom: H15+1
    | left-displaced-atom: H15+1
    | comments: "Methylated form of L-arginine"
    ]</pre>
        </div>

        <div class="large-5 cell notation">
          <h3>Atomic structures of inline monomeric forms</h3>
          <p>The <tt class="attribute">structure</tt> attribute describes the atomic structure of the monomeric form (excluding its backbone). This attribute must be a SMILES-encoded string. Each monomeric form can only have one structure attribute. This attribute is required to calculate the structure of the polymer.</p>
        </div>
        <div class="large-7 cell example">
          <h3>Example</h3>
          <div class="grid-x grid-padding-x">
            <div class="large-12 cell">
              <p>The text below illustrates how to describe the modified DNA nucleobase hypoxanthine, and the image below illustrates the molecule that the text specifies. The atom labels indicate the numbers of the atoms within the molecule. These numbers can be generated with OpenBabel <a href="http://openbabel.org/wiki/Category:Installation"><i class="fi-link"></i></a>.</p>
            </div>
            <div class="large-8 cell syntax">
              <pre>[id: "dI"
    | name: "hypoxanthine"
    | structure: "O=C1NC=NC2=C1N=CN2"
    ]</pre>
            </div>
            <div class="large-4 cell image">
              <img src="img/example/structure-dI.svg"/>
            </div>
          </div>
        </div>

        <div class="large-5 cell notation">
          <h3>Bonds between inline monomers and backbones</h3>
          <p>The <tt class="attribute">backbone-bond-atom</tt> and <tt class="attribute">backbone-displaced-atom</tt> attributes describe the bonds between inline monomeric forms and their backbones. The values of these attributes are the element of the atom and the position of the atom within the monomeric form (e.g. <tt>N3</tt>). OpenBabel can be used to display the numbers of the atoms within monomeric forms.</p>
          <p>Each monomeric form can form one or more bonds with the backbone and multiple atoms can be displaced. These attributes are required to calculate the structure of the polymer.</p>
        </div>
        <div class="large-7 cell example">
          <h3>Example</h3>
          <div class="grid-x grid-padding-x">
            <div class="large-12 cell">
              <p>The example below illustrates how to describe the modified DNA nucleotide deoxyinosine monophosphate. The red atoms in the image indicate the sugar-phosphate backbone.</p>
            </div>
            <div class="large-8 cell syntax">
              <pre>[id: "dI"
    | name: "hypoxanthine"
    | structure: "O=C1NC=NC2=C1N=CN2"
    | backbone-bond-atom: N10
    | backbone-displaced-atom: H10
    ]</pre>
            </div>
            <div class="large-4 cell image">
              <img src="img/example/monomer-backbone-bonds-dI.svg"/>
            </div>
          </div>
        </div>

        <div class="large-5 cell notation">
          <h3>Bonds between inline and adjacent monomers</h3>
          <p>The <tt class="attribute">left-bond-atom</tt>, <tt class="attribute">left-displaced-atom</tt>, <tt class="attribute">right-bond-atom</tt>, and <tt class="attribute">right-displaced-atom</tt> attributes describe the bonds between successive monomeric forms. The values of these attributes are the element of the atom and the position of the atom within the monomeric form (e.g. <tt>N3</tt>). OpenBabel can be used to display the numbers of the atoms within monomeric forms.</p>
          <p>Each monomeric form can have one or more bonds and displaced atoms with the preceding and following monomeric forms. These attributes are required to calculate the structure of the polymer.</p>
        </div>
        <div class="large-7 cell example">
          <h3>Example</h3>
          <div class="grid-x grid-padding-x">
            <div class="large-12 cell">
              <p>The example below illustrates how to describe the modified protein residue N<sup>5</sup>-methyl-L-arginine. The green atoms indicate the C terminus. The blue atoms indicate the N terminus.</p>
            </div>
            <div class="large-8 cell syntax">
              <pre>[id: "AA0305"
    | name: "N5-methyl-L-arginine"
    | structure: "O=C[C@H](CCCN(C(=[NH2])N)C)[NH3+]"
    | backbone-bond-atom: C2
    | backbone-displaced-atom: H2
    | right-bond-atom: C2
    | left-bond-atom: N15-1
    | left-displaced-atom: H15+1
    | left-displaced-atom: H15+1
    ]</pre>
            </div>
            <div class="large-4 cell image">
              <img src="img/example/left-right-bonds-AA0305.svg"/>
            </div>
            <div class="large-12 cell" style="margin-top:1rem;">
              <p>The example below illustrates how to describe a polymer with N<sup>5</sup>-methyl-L-arginine at the third position. The green lines indicate the bonds between the successive monomeric forms. The black labels indicate the positions of monomeric forms within the sequence.</p>
            </div>
            <div class="large-8 cell syntax">
              <pre>A[id: "AA0305"
    | name: "N5-methyl-L-arginine"
    | structure: "O=C[C@H](CCCN(C(=[NH2])N)C)[NH3+]"
    | backbone-bond-atom: C2
    | backbone-displaced-atom: H2
    | right-bond-atom: C2
    | left-bond-atom: N15-1
    | left-displaced-atom: H15+1
    | left-displaced-atom: H15+1
    ]A</pre>
            </div>
            <div class="large-4 cell image">
              <img src="img/example/left-right-bonds-A-AA0305-A.svg"/>
            </div>
          </div>
        </div>

        <div class="large-5 cell notation">
          <h3>Uncertainty about inline monomeric forms</h3>
          <p class="before_list">Through the inline monomeric forms, <i>BpForms</i> can represent two types of uncertainty in the primary sequences of forms of biopolymers:</p>
          <ul>
          <li>The <tt class="attribute">delta-mass</tt> <tt class="attribute">delta-charge</tt>  attributes can describe uncertainty in the atomic structure of the monomeric form.</li>
          <li>The <tt class="attribute">position</tt> attribute can describe uncertainty in the position of the monomeric form within the sequence.</i>
          </ul>
        </div>
        <div class="large-7 cell example">
          <h3>Examples</h3>
          <ul>
          <li><tt>[id: "dAMP" | delta-mass: 1 | delta-charge: 1]</tt>: indicates the presence of an additional proton whose exact location is not known.</li>
          <li><tt>[id: "dI" | position: 2-3]</tt>: indicates that deoxyinosine may occur anywhere between the second and third position.</li>
          </ul>
        </div>


        <div class="large-5 cell notation">
          <h3>Metadata about monomeric forms</h3>
          <p class="before_list"><i>BpForms</i> can represent several types of metadata about inline monomeric forms:</p>
          <ul>
            <li>The <tt class="attribute">id</tt> and <tt class="attribute">name</tt> attributes are human-readable labels for monomeric forms. The id attribute should be used to represent a short label. Only one id and one name is allowed per monomeric form.</li>
            <li>The <tt class="attribute">synonym</tt> attribute is an additional human-readable label. Monomeric forms can have multiple synonyms.</li>
            <li>The <tt class="attribute">identifier</tt> attribute indicates entries in namespaces (e.g., databases and ontologies) which are equivalent to the monomeric form. The value of this attribute should be an id within a namespace and the prefix of the namespaces as registered at Identifiers.org <a href="https://identifiers.org/"><i class="fi-link"></i></a> separated by "@" (e.g., <tt>"CHEBI:28997" @ "chebi"</tt>). Monomeric forms can have multiple identifiers.</li>
            <li>The <tt class="attribute">comments</tt> attribute can describe additional information about the monomeric form. Monomeric forms can only have one comment.</li>
          </ul>
        </div>
        <div class="large-7 cell example">
          <h3>Examples</h3>
          <ul>
          <li><tt>[id: "dI" | name: "deoxyinosine"]</tt>: represents the id and name of deoxyinosine.</li>
          <li><tt>[id: "dI" | synonym: "deoxyinosine" | synonym: "2'-deoxyinosine"]</tt>: represents multiple synonyms of deoxyinosine.</li>
          <li><tt>[id: "dI" | identifier: "CHEBI:28997" @ "chebi" | identifier: "65058" @ "pubchem.compound"]</tt>: represents equivalent entries in ChEBI and PubChem to deoxyinosine.</li>
          <li><tt>[id: "dI" | comments: "A purine 2'-deoxyribonucleoside that is inosine ..."]</tt>: represents comments about deoxyinosine.</li>
          </ul>
        </div>

        <div class="large-5 cell notation">
          <h3>Circular and linear topologies</h3>
          <p>By default, <i>BpForms</i> describes linear polymers. The <tt class="attribute">circular</tt> polymer attribute can be used to indicate that the polymer is circular (there is a bond between the last and first monomeric forms).</p>
        </div>
        <div class="large-7 cell example">
          <h3>Example</h3>
          <div class="grid-x grid-padding-x">
            <div class="large-12 cell">
              <p>The example below illustrates how to describe the circular DNA dimer of the DNA nucleotides deoxyadenosine monophosphate and dexocytosine monophosphate. The red atoms indicates the sugar-phosphate backbones. The green lines indicate the bonds between successive sugar-phosphate backbones. The black labels indicate the positions of monomeric forms within the sequence.</p>
            </div>
            <div class="large-8 cell syntax">
              <pre>AC | circular</pre>
            </div>
            <div class="large-4 cell image">
              <img src="img/example/circular-DNA-AC.svg"/>
            </div>
          </div>
        </div>

        <div class="large-5 cell notation">
          <h3>Crosslinks between non-adjacent monomers</h3>
          <p>The <tt class="attribute">crosslink</tt> polymer attribute can be used to indicate bonds between non-adjacent monomeric forms. For example, this attribute can be used to describe intrastrand crosslinks in DNA and disulfide bonds between cysteines in proteins. The values of these attributes are the position of the monomeric form within the sequence, the element of the atom, and the position of the atom within the monomeric form (e.g. <tt>8N3</tt>). OpenBabel can be used to display the numbers of the atoms within monomeric forms.</p>
          <p>Polymers can have zero, one, or more crosslinks.</p>
        </div>
        <div class="large-7 cell example">
          <h3>Example</h3>
          <div class="grid-x grid-padding-x">
            <div class="large-12 cell">
              <p>The example below illustrates how to describe a tripeptide with a disulfide bond. The blue line indicates the disulfide bond (crosslink). The green lines indicate the bonds between the successive residues. The black labels indicate the positions of monomeric forms within the sequence.</p>
            </div>
            <div class="large-8 cell syntax">
              <pre>CAC | crosslink: [
    left-bond-atom: 1S1 |
    left-displaced-atom: 1H1 |
    right-bond-atom: 3S1 |
    right-displaced-atom: 3H1
]</pre>
            </div>
            <div class="large-4 cell image">
              <img src="img/example/crosslink-protein-sulfide-bond.svg"/>
            </div>
          </div>
        </div>
      </div>

      <div class="grid-x grid-padding-x blue">
        <div class="large-12 cell">
          <h2>Resources for determining the sequences of biopolymer forms</h2>
          <p class="before_list">Below are several resources which can be helpful for determining the sequences of natural biopolymers and designing the sequences of synthetic biopolymers.</p>
        </div>
        <div class="large-6 cell">
          <h3>DNA</h3>
          <ul>
            <li><i>DNAMod</i> <a href="https://dnamod.hoffmanlab.org/"><i class="fi-link"></i></a>: Database of non-canonical DNA nucleobases</li>
            <li><i>MethDB</i> <a href="http://www.methdb.net/"><i class="fi-link"></i></a>: Database of non-canonical DNA</li>
            <li><i>MethSMRT</i> <a href="http://sysbio.gzzoc.com/methsmrt/"><i class="fi-link"></i></a>: Database of non-canonical DNA</li>
            <li><i>REPAIRtoire</i> <a href="http://repairtoire.genesilico.pl/damage/"><i class="fi-link"></i></a>: Database of DNA damages</li>
          </ul>

          <h3>RNA</h3>
          <ul>
            <li><i>MODOMICS</i> <a href="http://modomics.genesilico.pl/modifications/"><i class="fi-link"></i></a>: Database of non-canonical RNA nucleosides</li>
            <li><i>RMBase</i> <a href="http://rna.sysu.edu.cn/rmbase/"><i class="fi-link"></i></a>: Database of modified RNA</li>
            <li><i>RNA Modification Database</i> <a href="https://mods.rna.albany.edu/"><i class="fi-link"></i></a>: Database of modified RNA</li>
          </ul>

          <h3>Drawing structures of monomeric forms</h3>
          <ul>
            <li><i>ChemAxon Marvin</i> <a href="https://chemaxon.com/products/marvin"><i class="fi-link"></i></a>: Software for drawing structures of monomeric forms</li>
            <li><i>OpenBabel</i> <a href="http://openbabel.org/wiki/Category:Installation"><i class="fi-link"></i></a>: Software for calculating the numbers of the atoms in monomeric forms</li>
          </ul>
        </div>
        <div class="large-6 cell">
          <h3>Proteins</h3>
          <ul>
            <li><i>dbPTM</i> <a href="http://dbptm.mbc.nctu.edu.tw/"><i class="fi-link"></i></a>: Database of non-canonical amino acids</li>
            <li><i>Delta Mass</i> <a href="https://abrf.org/delta-mass"><i class="fi-link"></i></a>: Database of modified amino acids</li>
            <li><i>FindMod</i> <a href="https://web.expasy.org/findmod/findmod_masses.html"><i class="fi-link"></i></a>: Database of post-translational modifications</li>
            <li><i>iPTMnet</i> <a href="https://research.bioinformatics.udel.edu/iptmnet"><i class="fi-link"></i></a>: Database of post-translational modifications</li>
            <li><i>ProForma</i> <a href="https://www.topdownproteomics.org/resources/proforma/"><i class="fi-link"></i></a>: Notation for protein forms.  Note, this notation is not unambiguous. This limits its abiltiy to facilitate data integration and the calculation of properties of protein forms.</li>
            <li><i>PDB Chemical Components</i> <a href="http://www.wwpdb.org/data/ccd"><i class="fi-link"></i></a>: Database of modified amino acids</li>
            <li><i>PDB in Europe Chemical Components</i> <a href="https://www.ebi.ac.uk/pdbe/pdb-component-library/index.html>"><i class="fi-link"></i></a>: Database of modified amino acids</li>
            <li><i>PhosphoSitePlus</i> <a href="https://www.phosphosite.org"><i class="fi-link"></i></a>: Database of protein phosphorylations</li>
            <li><i>Protein Ontology</i> <a href="https://pir.georgetown.edu/pro/"><i class="fi-link"></i></a>: Database of modified proteins</li>
            <li><i>PSIMOD</i> <a href="https://bioportal.bioontology.org/ontologies/PSIMOD"><i class="fi-link"></i></a>: Ontology of non-canonical amino acids</li>
            <li><i>RESID</i> <a href="https://pir.georgetown.edu/resid/"><i class="fi-link"></i></a>: Database of non-canonical protein residues</li>
            <li><i>UniMod</i> <a href="http://www.unimod.org/"><i class="fi-link"></i></a>: Database of non-canonical amino acids</li>
            <li><i>UniProt</i> <a href="https://www.uniprot.org/"><i class="fi-link"></i></a>: Database of modified amino acids in proteins</li>
            <li><i>UniProt Controlled Vocabulary of Posttranslational Modifications</i> <a href="https://www.uniprot.org/docs/ptmlist"><i class="fi-link"></i></a>: Database of modified amino acids</li>
          </ul>
        </div>
      </div>

      <div class="grid-x grid-padding-x orange">
        <div class="large-12 cell">
          <h2>Uses cases for epigenomics, proteomics, systems biology, synthetic biology, and proteomics</h2>
          <p class="before_list">By concretely capturing the primary structure of biopolymers, <i>BpForms</i> can help facilitate a wide range of epigenomics, proteomics, proteomics systems biology, and synthetic biology research.</p>
        </div>

        <div class="large-4 cell">
          <a name="epigenomics"></a>
          <h3>Epigenomics</h3>
          <p><i>BpForms</i> can help researchers precisely communicate the structures of modified DNA, such as methylations, that bacteria use to distinguish self from non-self DNA. We anticipate that this will be increasingly important as researchers continue to discover new types of modifications and begin to investigate their impact on the interactions of proteins with DNA.</p>
        </div>
        <div class="large-8 cell">
          <h3>Example</h3>
          <p>Several chemotherapeutics, such as cisplatin, cause toxic side effects by damaging the DNA of healthy cells. Cells have several pathways with overlapping functions to repair DNA damage. This includes direct repair, base excision repair, nucleotide excision repair, and homologous recombination. Because chemotherapeutics cause a wide range of damage and because cells have several pathways to repair DNA damage, it is challenging to assemble an integrated understanding of the repair of DNA damage caused by chemotherapeutics. <i>BpForms</i> could help researchers develop an integrated understanding of DNA repair by helping researchers concretely communicate the damage caused by each chemotherapeutic and the types of damage repaired by each pathway.</p>
        </div>

        <div class="large-4 cell">
          <a name="proteomics"></a>
          <h3>Proteomics</h3>
          <p><i>BpForms</i> can help researchers precisely communicate the sequences of proteoforms, analyze modifications, and improve the quality of reported sequences by identifying errors in the descriptions of proteoforms, such as monomeric forms that are inconsistent with the unmodified sequence (e.g., selenocysteine modification of a non-cysteine residue) and inconsistent bonds (e.g., N,N,N-trimethyl-L-alanine, which has not N-terminus, located in the middle of a peptide).</p>
        </div>
        <div class="large-8 cell">
          <h3>Example: Analysis of the metabolic load of protein modification in humans</h3>
          <p>PRO contains curated modifications of 2,312 human proteins <a href="https://proconsortium.org/"><i class="fi-link"></i></a>. We used <i>BpForms</i> together with PRO to assess the metabolic cost of protein modification in humans. We found that human proteins have 1.7 &plusmn;1.6 modifications that increase their mass by 146.4 &plusmn; 154.3 Da and decrease their charge by 3.05 &plusmn; 3.30. This analysis also led us to improve PRO by correcting four types of errors in the curated modifications. The code is available at GitHub <a href="https://github.com/KarrLab/bpforms/blob/master/examples/pro.py"><i class="fi-link"></i></a>.</p>
          <!-- 
          <ol>
            <li>Identified 1 entry with a obsolete UniProt ids.</li>
            <li>Identified 3 entries with processing positions that are longer than the length of the sequence.</li>
            <li>Identified 58 entries with modifications whose positions and residues are inconsistent.</li>
            <li>Identified 4 entries with multiple modifications annotated for the same position.</li>
          </ol>
          -->

          <img src="img/example/pro_nonreasoned.svg" style="margin-bottom: 0.5rem;" />
        </div>

        <div class="large-6 cell">
          <a name="sys-bio"></a>
          <h3>Systems biology</h3>
          <p><i>BpForms</i> can help modelers describe the semantic meaning of models by precisely describing the species in models. Importantly, this precision makes it easier for other researchers to understand, reuse, extend, and compose models for other studies. <i>BpForms</i> can also help modelers build more comprehensive models by helping identify gaps in models such as missing intermediate modification states of proteins and missing interactions between modification states. In particular, <i>BpForms</i> can help modelers identify the full combinatorial complexity of biochemistry that should be modeled. In addition, <i>BpForms</i> can help researchers increase the quality of models by helping identify errors such as element imbalances.</p>
        </div>
        <div class="large-6 cell">
          <h3>Example</h3>
          <p>The Kholodenko model of the eukaryotic MAPK signaling cascade (DOI: 10.1046/j.1432-1327.2000.01197.x <a href="https://doi.org/10.1046/j.1432-1327.2000.01197.x"><i class="fi-link"></i></a>, BioModels: BIOMD0000000010 <a href="https://www.ebi.ac.uk/biomodels/BIOMD0000000010"><i class="fi-link"></i></a>) represents the biphosporylation of Mek1/MAPKK by Mos/MAPKKK and the biphosporylation of Erk2/MAPK by Mek1/MAPKK. Annotating the structures of the species in the model with <i>BpForms</i>, enabled us to identify two gaps in the model: two additional intermediate phosphorylation forms of Mek1 and Erk2 and the reactions involving these species. <i>BpForms</i> also enabled us to identify several unbalanced reactions that do not capture phosphate donors.</p>
        </div>

        <div class="large-4 cell">
          <a name="syn-bio"></a>
          <h3>Synthetic biology</h3>
          <p><i>BpForms</i> can help engineers precisely represent and communicate the structures of parts for synthetic organisms. In addition, <i>BpForms</i> could help engineers identify the dependencies and interfaces of parts which, in turn, could help engineers use parts in alternative hosts, compose parts, and share parts more reliably.</p>
        </div>
        <div class="large-8 cell">
          <h3>Example</h3>
          <p><i>E. coli</i> pyruvate dehydrogenase requires lipoate ligation at L43 of the active site of the E1 subunit (UniProt: P96104 <a href="https://www.uniprot.org/uniprot/P96104"><i class="fi-link"></i></a>). By representing lipoate ligation of the E1 subunit, <i>BpForms</i> can help capture the dependence of <i>E. coli</i> pyruvate dehydrogenase on lipoate ligase. In turn, this could help engineers recognize that <i>E. coli</i> pyruvate dehydrogenase can only be used in other hosts that have a lipoate ligase, or that a lipoate ligase, such as LplA (UnitProt: P32099 <a href="https://www.uniprot.org/uniprot/P32099"><i class="fi-link"></i></a>), must be co-transformed with <i>E. coli</i> pyruvate dehydrogenase.</p>
        </div>

        <div class="large-4 cell">
          <a name="transcriptomics"></a>
          <h3>Transcriptomics</h3>
          <p><i>BpForms</i> can help researchers precisely communicate the sequences of rRNA, tRNA, and other non-coding RNA; analyze RNA modifications; and improve the quality of reported sequences by identifying errors in the descriptions of modified RNA such as undefined monomeric forms and inconsistent bonds (e.g., 3' caps that are not located at the 3' position).
          </p>
        </div>
        <div class="large-8 cell">
          <h3>Example: Analysis of the metabolic load of rRNA and tRNA modification in <i>E. coli</i></h3>
          <p>MODOMICS contains 732 curated sequences of rRNA and tRNA <a href="http://modomics.genesilico.pl/sequences/"><i class="fi-link"></i></a>. We used <i>BpForms</i> together with MODOMICS to assess the metabolic cost of RNA modification in <i>Escherichia coli</i>. We found that <i>E. coli</i> tRNA have 7.8 &plusmn; 2.2 modifications that increase their mass by 177.5 &plusmn; 108.0 Da and charge by 0.62 &plusmn; 0.76. This analysis also led us to add missing information about the origin of several of the monomeric forms derived from MODOMICS and correct three types of errors in the MODOMICS RNA sequences. The code is available at GitHub <a href="https://github.com/KarrLab/bpforms/blob/master/examples/modomics.py"><i class="fi-link"></i></a>.</p>
          <!-- <ol>
            <li>We added missing information about origin of several monomeric forms.</li>
            <li>We corrected three types of errors in the sequences. (a) Lower case codes which do not correspond to a monomeric form, and which could be confused with descriptions of repetitive elements. (b) Trailing spaces which could be confused with monomeric forms. (c) Underscores used to represent insertions that were not described as a part of the MODOMICS syntax, and which could be confused with monomeric forms.</li>
            <li>involving 7.8 &plusmn; 4.4 carbon atoms,
              18.3 &plusmn; 8.6 hydrogen atoms, 1.9 &plusmn; 1.2 nitrogen atoms, 1.8 &plusmn; 2.5 oxygen atoms, and 1.7 &plusmn; 0.7 sulfur atoms.</li>
          </ol> -->

          <img src="img/example/modomics.trna.code-freq.svg" style="margin-bottom: 0.5rem;" />
        </div>
      </div>

      <div class="grid-x grid-padding-x blue">
        <div class="large-12 cell">
          <h2>Integrating <i>BpForms</i> into the BioPAX, CellML, FASTA, PDB, SBML, and SBOL standards</h2>
          <p class="before_list"><i>BpForms</i> can be used in conjunction with several commonly used standards in genomics, transcriptomics, proteomics, systems biology, and synthetic biology. In addition, <i>BpForms</i> can easily be embedded within other documents such as Excel workbooks and comma-separated tables.</p>
        </div>

        <div class="large-4 cell">
          <a name="biopax"></a>
          <h3>BioPAX: Knowledge of pathways</h3>
          <p><i>BpForms</i> can be used to concretely describe all of the DNA, RNA, and proteins involved in pathways encoded in BioPAX <a href="https://www.biopax.org/"><i class="fi-link"></i></a>. <i>BpForms</i> can be used with the <tt>sequence</tt> child of <tt>DNAReference</tt>, <tt>RNAReference</tt>, and <tt>ProteinReference</tt> objects.</p>
        </div>
        <div class="large-8 cell example">
          <h3>Examples</h3>
          <p style="margin-bottom: 0.5rem;"><b>DNA:</b> <i>E. coli</i> K-12 MG1655 Dam 6-methyladenine sites (701..914) involved in host recognition <a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_000913.3"><i class="fi-link"></i></a></p>
          <pre class="small">...
  &lt;bp:DNA&gt;
    &lt;bp:entityReference&gt;
      &lt;bp:DNAReference&gt;
        &lt;bp:sequence
          rdf:datatype="http://www.w3.org/2001/XMLSchema#string"
          rdf:about="http://edamontology.org/format_3909#dna"&gt;
          <span class="bpform">...
          TGATTTGCCGTGGCGAGAAAATGTCG<span class="modification">{a}</span>TCGCCATTATGGCCGGCGTATTAGAAGCGCGCGGTCACAAC
          GTTACTGTTATCG<span class="modification">{a}</span>TCCGGTCGAAAAACTGCTGGCAGTGGGGCATTACCTCGAATCTACCGTCGATAT
          TGCTGAGTCCACCCGCCGTATTGCGGCAAGCCGCATTCCGGCTG<span class="modification">{a}</span>TCACATGGTGCTGATGGCAGGTT
          ...</span>
        &lt;/bp:sequence&gt;
      &lt;/bp:DNAReference&gt;
    &lt;/bp:entityReference&gt;
  &lt;/bp:DNA&gt;
...
</pre>

          <p style="margin-top: 1rem; margin-bottom: 0.5rem;"><b>RNA:</b> Modifications of <i>B. subtilis</i> tRNA<sup>UGC</sup> involved in stability <a href="http://modomics.genesilico.pl/sequences/list/?type_field=tRNA&subtype=Ala&species=4&display=Display&nomenclature=abbrev"><i class="fi-link"></i></a></p>
          <pre class="small">...
  &lt;bp:RNA&gt;
    &lt;bp:entityReference&gt;
      &lt;bp:RNAReference&gt;
        &lt;bp:sequence
          rdf:datatype="http://www.w3.org/2001/XMLSchema#string"
          rdf:about="http://edamontology.org/format_3909#rna"&gt;
          <span class="bpform">GGAGCCUUAGCUCAGC<span class="modification">{8U}</span>GGGAGAGCGCCUGCUU<span class="modification">{501U}</span>GC<span class="modification">{6A}</span>CGCAGGAG<span class="modification">{7G}</span>UCAGCGG<span class="modification">{5U}</span><span class="modification">{9U}</span>CGAUCCCGCUAGGCUCCA
          CCA</span>
        &lt;/bp:sequence&gt;
      &lt;/bp:RNAReference&gt;
    &lt;/bp:entityReference&gt;
  &lt;/bp:RNA&gt;
...</pre>

          <p style="margin-top: 1rem; margin-bottom: 0.5rem;"><b>Protein:</b> Modifications of <i>H. sapiens</i> MAPK3 involved in signaling <a href="https://www.uniprot.org/uniprot/P27361#ptm_processing"><i class="fi-link"></i></a></p>
          <pre class="small">...
  &lt;bp:Protein&gt;
    &lt;bp:entityReference&gt;
      &lt;bp:ProteinReference&gt;
        &lt;bp:sequence
          rdf:datatype="http://www.w3.org/2001/XMLSchema#string"
          rdf:about="http://edamontology.org/format_3909#protein"&gt;
          <span class="bpform">M<span class="modification">{AA0041}</span>AAAAQGGGGGEPRRTEGVGPGVPGEVEMVKGQPFDVGPRYTQLQYIGEGAYGMVSSAYDHVRKTRVAIKKISPFEHQTYCQRTL
          REIQILLRFRHENVIGIRDILRASTLEAMRDVYIVQDLMETDLYKLLKSQQLSNDHICYFLYQILRGLKYIHSANVLHRDLKPSNLLINTTCD
          LKICDFGLARIADPEHDH<span class="modification">{AA0038}</span>GFL<span class="modification">{AA0038}</span>E<span class="modification">{AA0039}</span>VA<span class="modification">{AA0038}</span>RWYRAPEIMLNSKGYTKSIDIWSVGCILAEMLSNRPI
          FPGKHYLDQLNHILGILGSPSQEDLNCIINMKARNYLQSLPSKTKVAWAKLFPKSDSKALDLLDRMLTFNPNKRITVEEALAHPYLEQYYDPT
          DEPVAEEPFTFAMELDDLPKERLKELIFQETARFQPGVLEAP</span>
        &lt;/bp:sequence&gt;
      &lt;/bp:ProteinReference&gt;
    &lt;/bp:entityReference&gt;
  &lt;/bp:Protein&gt;
...</pre>
        </div>

        <div class="large-4 cell">
          <a name="cellml"></a>
          <h3>CellML: Kinetic models</h3>
          <p style="margin-bottom: 0.5rem;"><i>BpForms</i> can be used to concretely describe the meaning of each component of a model encoded in CellML <a href="https://www.cellml.org/"><i class="fi-link"></i></a>. <i>BpForms</i> can be used with the <tt class="attribute">RDF</tt> element of <tt class="attribute">component</tt> objects.</p>
        </div>
        <div class="large-8 cell example">
          <h3>Example</h3>
          <p style="margin-bottom: 0.5rem;"><b>Protein:</b> Phosphorylated Erk and Mek in signal transduction (DOI: 10.1038/msb.2009.4 <a href="https://doi.org/10.1038/msb.2009.4"><i class="fi-link"></i></a>, Physiome Model Repository <a href="https://models.physiomeproject.org/e/31"><i class="fi-link"></i></a>). See complete CellML file <a href="https://github.com/KarrLab/bpforms/blob/master/examples/wang_cirit_haugh_2009.cellml"><i class="fi-link"></i></a>.</p>

<pre class="small">...
  &lt;component cmeta:id="ypp" name="ypp"&gt;
      &lt;rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"&gt;
        &lt;rdf:Description rdf:about="#ypp"&gt;
          &lt;bpforms:seq xmlns:bpforms="https://bpforms.org/protein"&gt;
            <span class="bpform">MPKKKPTPIQLNPAPDGSAVNGTSSAETNLEALQKKLEELELDEQQRKRLEAFLTQKQKVGELKDDDFEKISELGAGNGGVVFKVSHKPSG
            LVMARKLIHLEIKPAIRNQIIRELQVLHECNSPYIVGFYGAFYSDGEISICMEHMDGGSLDQVLKKAGRIPEQILGKVSIAVIKGLTYLRE
            KHKIMHRDVKPSNILVNSRGEIKLCDFGVSGQLID<span class="modification">{AA0037}</span>MAN<span class="modification">{AA0037}</span>FVGTRSYMSPERLQGTHYSVQSDIWSMGLSLVEMAVG
            RYPIPPPDAKELELMFGCQVEGDAAETPPRPRTPGRPLSSYGMDSRPPMAIFELLDYIVNEPPPKLPSGVFSLEFQDFVNKCLIKNPAERA
            DLKQLMVHAFIKRSDAEEVDFAGWLCSTIGLNQPSTPTHAAGV</span>
          &lt;/bpforms:seq&gt;
        &lt;/rdf:Description&gt;
      &lt;/rdf:RDF&gt;
    &lt;/component&gt;
...</pre>
        </div>

        <div class="large-4 cell">
          <a name="fasta"></a>
          <h3>FASTA: Sets of sequences</h3>
          <p>Sets of <i>BpForms</i> can be encoded in FASTA files. Such files can be written with the <tt>bpforms.util.write_to_fasta</tt> function or packages such as BioPython <a href="https://biopython.org/wiki/SeqIO"><i class="fi-link"></i></a>.</p>
        </div>
        <div class="large-8 cell example">
          <h3>Example</h3>
          <p style="margin-bottom: 0.5rem;"><b>Protein:</b> multiple phosphorylated forms of <i>H. sapiens</i> MAPK <a href="https://www.uniprot.org/uniprot/Q02750"><i class="fi-link"></i></a></p>
          <pre class="small">&gt; y | MEK | Q02750
<span class="bpform">MPKKKPTPIQLNPAPDGSAVNGTSSAETNLEALQKKLEELELDEQQRKRLEAFLTQKQKVGELKDDDFEKISELGAGNGGVVFKVSHKPSGLVMARKLIH
LEIKPAIRNQIIRELQVLHECNSPYIVGFYGAFYSDGEISICMEHMDGGSLDQVLKKAGRIPEQILGKVSIAVIKGLTYLREKHKIMHRDVKPSNILVNS
RGEIKLCDFGVSGQLIDSMANSFVGTRSYMSPERLQGTHYSVQSDIWSMGLSLVEMAVGRYPIPPPDAKELELMFGCQVEGDAAETPPRPRTPGRPLSSY
GMDSRPPMAIFELLDYIVNEPPPKLPSGVFSLEFQDFVNKCLIKNPAERADLKQLMVHAFIKRSDAEEVDFAGWLCSTIGLNQPSTPTHAAGV</span>

&gt; yp | phosphorylated MEK | Q02750 | pS218
<span class="bpform">MPKKKPTPIQLNPAPDGSAVNGTSSAETNLEALQKKLEELELDEQQRKRLEAFLTQKQKVGELKDDDFEKISELGAGNGGVVFKVSHKPSGLVMARKLIH
LEIKPAIRNQIIRELQVLHECNSPYIVGFYGAFYSDGEISICMEHMDGGSLDQVLKKAGRIPEQILGKVSIAVIKGLTYLREKHKIMHRDVKPSNILVNS
RGEIKLCDFGVSGQLID<span class="modification">{AA0037}</span>MANSFVGTRSYMSPERLQGTHYSVQSDIWSMGLSLVEMAVGRYPIPPPDAKELELMFGCQVEGDAAETPPRPRTP
GRPLSSYGMDSRPPMAIFELLDYIVNEPPPKLPSGVFSLEFQDFVNKCLIKNPAERADLKQLMVHAFIKRSDAEEVDFAGWLCSTIGLNQPSTPTHAAGV</span></pre>
        </div>

        <div class="large-4 cell">
          <a name="pdb"></a>
          <h3>PDB: Macromolecule 3D structures</h3>
          <p><i>BpForms</i> can be used to provide human readable annotations of protein structures encoded in PDB files <a href="http://www.wwpdb.org/documentation/file-format"><i class="fi-link"></i></a>. <i>BpForms</i> can be embedded within <tt>REMARK</tt> records.
          </p>
        </div>
        <div class="large-8 cell example">
          <h3>Example</h3>
          <p style="margin-bottom: 0.5rem;"><b>Protein:</b> <i>Bos taurus</i> selenocysteine synthase Gpx1 <a href="https://www.rcsb.org/structure/1gp1"><i class="fi-link"></i></a></p>
          <pre class="small">...
REMARK   1                                                                      
REMARK   1 >1GP1:A
REMARK   1 <span class="bpform">AAALAAAAPRTVYAFSARPLAGGEPFNLSSLRGKVLLIENVASL<span class="modification">{SE7}</span>GTTVRDYTQMNDLQRRLGPRGLVVLGFPCNQFGHQENAKNEEIL</span>
REMARK   1 <span class="bpform">NCLKYVRPGGGFEPNFMLFEKCEVNGEKAHPLFAFLREVLPTPSDDATALMTDPKFITWSPVCRNDVSWNFEKFLVGPDGVPVRRYSRRFLTI</span>
REMARK   1 <span class="bpform">DIEPDIETLLSQGASA</span>
REMARK   1 >1GP1:B
REMARK   1 <span class="bpform">AAALAAAAPRTVYAFSARPLAGGEPFNLSSLRGKVLLIENVASL<span class="modification">{SE7}</span>GTTVRDYTQMNDLQRRLGPRGLVVLGFPCNQFGHQENAKNEEIV</span>
REMARK   1 <span class="bpform">VLGFPCNQFGHQENAKNEEILNCLKYVRPGGGFEPNFMLFEKCEVNGEKAHPLFAFLREVLPTPSDDATALMTDPKFITWSPVCRNDVSWNFE</span>
REMARK   1 <span class="bpform">KFLVGPDGVPVRRYSRRFLTIDIEPDIETLLSQGASA</span>
...</pre>
        </div>

        <div class="large-4 cell">
          <a name="sbml"></a>
          <h3>SBML: Kinetic models</h3>
          <p><i>BpForms</i> can be used to concretely describe the meaning of each species in a model encoded in Systems Biology Markup Language (SBML) <a href="https://sbml.org"><i class="fi-link"></i></a>. <i>BpForms</i> can be used with the <tt class="attribute">annotation</tt> element of <tt class="attribute">species</tt> objects.</p>
        </div>
        <div class="large-8 cell example">
          <h3>Examples</h3>

          <p style="margin-bottom: 0.5rem;"><b>Protein:</b> Phosphorylated Cdc2 and Cdc12 in the yeast cell cycle (DOI: 10.1073/pnas.88.16.7328 <a href="https://doi.org/10.1073/pnas.88.16.7328"><i class="fi-link"></i></a>, BioModels: BIOMD0000000005 <a href="https://www.ebi.ac.uk/biomodels/BIOMD0000000005"><i class="fi-link"></i></a>). See complete SBML file <a href="https://github.com/KarrLab/bpforms/blob/master/examples/BIOMD0000000005.xml"><i class="fi-link"></i></a>.</p>
          <pre class="small">...
  &lt;species name="cdc2k-p" metaid="cdc2k"&gt;
    &lt;annotation&gt;
      &lt;rdf:RDF
        xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"&gt;
        &lt;rdf:Description rdf:about="#cdc2k-p"&gt;
          &lt;bpforms:seq xmlns:bpforms="https://bpforms.org/protein"&gt;
            <span class="bpform">MENYQKVEKIGEG<span class="modification">{AA0038}</span>span&gt;<span class="modification">{AA0039}</span>GVVYKARHKLSGRIVAMKKIRLEDESEGVPSTAIREISLLKEVNDENNRSNCVRLLDI
            LHAESKLYLVFEFLDMDLKKYMDRISETGATSLDPRLVQKFTYQLVNGVNFCHSRRIIHRDLKPQNLLIDKEGNLKLADFGLARSFGVPLRN
            Y<span class="modification">{AA0038}</span>HEIVTLWYRAPEVLLGSRHYSTGVDIWSVGCIFAEMIRRSPLFPGDSEIDEIFKIFQVLGTPNEEVWPGVTLLQDYKSTFPRW
            KRMDLHKVVPNGEEDAIELLSAMLVYDPAHRISAKRALQQNYLRDFH</span>
          &lt;/bpforms:seq&gt;
        &lt;/rdf:Description&gt;
      &lt;/rdf:RDF&gt;
    &lt;/annotation&gt;
  &lt;/species&gt;
...</pre>

          <p style="margin-top: 1rem; margin-bottom: 0.5rem;"><b>Protein:</b> Phosphorylated Mos/Raf1, Mek1, and Erk2 in the eukaryote MAPK cascade (DOI: 10.1046/j.1432-1327.2000.01197.x <a href="https://doi.org/10.1046/j.1432-1327.2000.01197.x"><i class="fi-link"></i></a>, BioModels: BIOMD0000000010 <a href="https://www.ebi.ac.uk/biomodels/BIOMD0000000010"><i class="fi-link"></i></a>). See complete SBML file <a href="https://github.com/KarrLab/bpforms/blob/master/examples/BIOMD0000000010.xml"><i class="fi-link"></i></a>.</p>
          <pre class="small">...
  &lt;species name="Erk2-PP" metaid="_584615"&gt;
    &lt;annotation&gt;
      &lt;rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"&gt;
        &lt;rdf:Description rdf:about="#_584615"&gt;
          &lt;bpforms:seq xmlns:bpforms="https://bpforms.org/protein"&gt;
            <span class="bpform">MAAAGAASNPGGGPEMVRGQAFDVGPRYINLAYIGEGAYGMVCSAHDNVNKVRVAIKKISPFEHQTYCQRTLREIKILLRFKHENIIGINDI
            IRAPTIEQMKDVYIVQDLMETDLYKLLKTQHLSNDHICYFLYQILRGLKYIHSANVLHRDLKPSNLLLNTTCDLKICDFGLARVADPDHDHT
            GFL<span class="modification">{AA0038}</span>E<span class="modification">{AA0039}</span>VATRWYRAPEIMLNSKGYTKSIDIWSVGCILAEMLSNRPIFPGKHYLDQLNHILGILGSPSQEDLNCIINLK
            ARNYLLSLPHKNKVPWNRLFPNADPKALDLLDKMLTFNPHKRIEVEAALAHPYLEQYYDPSDEPVAEAPFKFEMELDDLPKETLKELIFEET
            ARFQPGY</span>
          &lt;/bpforms:seq&gt;
        &lt;/rdf:Description&gt;
      &lt;/rdf:RDF&gt;
    &lt;/annotation&gt;
  &lt;/species&gt;
...</pre>
        </div>

        <div class="large-4 cell">
          <a name="sbol"></a>
          <h3>SBOL: Designs for synthetic parts</h3>
          <p><i>BpForms</i> can be used to describe the meaning of each DNA, RNA, and protein molecule in genetic designs encoded in the Synthetic Biology Open Language (SBOL) <a href="http://sbolstandard.org/"><i class="fi-link"></i></a>. <i>BpForms</i> can be used with the <tt>elements</tt> attribute of <tt>Sequence</tt> objects.</p>
          <p style="margin-bottom: 0rem;">The following URIs should be used to indicate the encodings for the sequences of DNA, RNA, and protein molecules.</p>
          <ul>
            <li>DNA: http://edamontology.org/format_3909#dna</li>
            <li>RNA: http://edamontology.org/format_3909#rna</li>
            <li>Protein: http://edamontology.org/format_3909#protein</li>
          </ul>
          <p>See SBOL SEP 033 <a href="https://github.com/SynBioDex/SEPs/blob/master/sep_033.md"><i class="fi-link"></i></a> and issue 77 <a href="https://github.com/SynBioDex/SEPs/issues/77"><i class="fi-link"></i></a> for more information.</p>
        </div>
        <div class="large-8 cell example">
          <h3>Examples</h3>

          <p style="margin-bottom: 0.5rem;"><b>RNA:</b> Modified <i>B. subtilis</i> tRNA<sup>ILE</sup> 69 (SynBioHub: BO_28687 <a href="https://synbiohub.org/public/bsu/BO_28687"><i class="fi-link"></i></a>). See complete SBOL file <a href="https://github.com/KarrLab/bpforms/blob/master/examples/BO_28687.xml"><i class="fi-link"></i></a>.</p>
          <pre class="small">...
  &lt;sbol:Sequence&gt;
    &lt;sbol:elements>
      <span class="bpform">GGGCCUGUAGCUCAGC<span class="modification">{8U}</span>GG<span class="modification">{8U}</span><span class="modification">{8U}</span>AGAGCGCACGCCUGAU<span class="modification">{62A}</span>AGCGUGAG<span class="modification">{7G}</span>UCGAUGG<span class="modification">{5U}</span><span class="modification">{9U}</span>CGAGUCCAUUCAGGCCCACCA</span>
    &lt;/sbol:elements&gt;
    &lt;sbol:encoding rdf:resource="http://edamontology.org/format_3909#rna"/&gt;
  &lt;/sbol:Sequence&gt;
...</pre>

          <p style="margin-top: 1rem; margin-bottom: 0.5rem;"><b>Protein:</b> Lipoate-ligated acetyltransferase component PdhC of <i>B. subtilis</i> pyruvate dehydrogenase complex (SynBioHub: BO_32431 <a href="https://synbiohub.org/public/bsu/BO_32431"><i class="fi-link"></i></a>). See complete SBOL file <a href="https://github.com/KarrLab/bpforms/blob/master/examples/BO_32431.xml"><i class="fi-link"></i></a>.</p>
          <pre class="small">...
  &lt;sbol:Sequence&gt;
    &lt;sbol:elements>
      <span class="bpform">MAFEFKLPDIGEGIHEGEIVKWFVKPNDEVDEDDVLAEVQND<span class="modification">{AA0118}</span>AVVEIPSPVKGKVLELKVEEGTVATVGQTIITFDAPGYEDLQFKGSDE
      SDDAKTEAQVQSTAEAGQDVAKEEQAQEPAKATGAGQQDQAEVDPNKRVIAMPSVRKYAREKGVDIRKVTGSGNNGRVVKEDIDSFVNGGAQEAAPQE
      TAAPQETAAKPAAAPAPEGEFPETREKMSGIRKAIAKAMVNSKHTAPHVTLMDEVDVTNLVAHRKQFKQVAADQGIKLTYLPYVVKALTSALKKFPVL
      NTSIDDKTDEVIQKHYFNIGIAADTEKGLLVPVVKNADRKSVFEISDEINGLATKAREGKLAPAEMKGASCTITNIGSAGGQWFTPVINHPEVAILGI
      GRIAEKAIVRDGEIVAAPVLALSLSFDHRMIDGATAQNALNHIKRLLNDPQLILMEA</span>
    &lt;/sbol:elements&gt;
    &lt;sbol:encoding rdf:resource="http://edamontology.org/format_3909#protein"/&gt;
  &lt;/sbol:Sequence&gt;
...</pre>
        </div>
      </div>

      <div class="grid-x grid-padding-x orange">
        <div class="large-12 cell">
          <h2><i>BpForm</i> interfaces</h2>
          <p class="before_list"><i>BpForms</i> includes four software interfaces for verifying descriptions of biopolymers and calculating properties such as their atomic structures, formulae, molecular weights, and charges.</p>
        </div>
        <div class="large-3 cell">
          <h3>Webform</h3>
          The webform above <a href="#form"><i class="fi-link"></i></a> can be used to validate <i>BpForms</i> and calculate their properties.
        </div>
        <div class="large-3 cell">
          <h3>JSON REST API</h3>
          A JSON REST API is available at <a href="https://bpforms.org/api">https://bpforms.org/api</a>.
        </div>
        <div class="large-3 cell">
          <h3>Command line interface</h3>
          A command line interface is available from PyPI <a href="https://pypi.python.org/pypi/bpforms"><i class="fi-link"></i></a>.
        </div>
        <div class="large-3 cell">
          <h3>Python library</h3>
          A Python library is available from PyPI <a href="https://pypi.python.org/pypi/bpforms"><i class="fi-link"></i></a>.
        </div>
      </div>

      <div class="grid-x grid-padding-x blue">
        <div class="large-12 cell">
          <h2>Tutorials, documentation, and help</h2>
        </div>
        <div class="large-6 cell">
          <h3>Documentation for the notation</h3>
          Documentation for the notation is available above <a href="#notation"><i class="fi-link"></i></a>. The grammar for the notation is available at GitHub <a href="https://raw.githubusercontent.com/KarrLab/bpforms/master/bpforms/grammar.lark"><i class="fi-link"></i></a>
        </div>
        <div class="large-6 cell">
          <h3>Definitions of the monomeric forms in the alphabets</h3>
          Documentation for the alphabets is available above <a href="#alphabets"><i class="fi-link"></i></a>. This includes images and detailed information about each monomeric form in each alphabet.
        </div>
        <div class="large-6 cell">
          <h3>Query builder for the REST API</h3>
          A visual interface for building REST queries is available at bpforms.org/api <a href="https://www.bpforms.org/api"><i class="fi-link"></i></a>.
        </div>
        <div class="large-6 cell">
          <h3>Documentation for the REST API</h3>
          Documentation for the REST API is available at bpforms.org/api <a href="https://www.bpforms.org/api"><i class="fi-link"></i></a>.
        </div>
        <div class="large-6 cell">
          <h3>Documentation for the command line program</h3>
          Documentation for the command line program is available inline by running <tt>bpforms --help</tt>.
        </div>
        <div class="large-6 cell">
          <h3>Tutorial for the Python API</h3>
          A Jupyter notebook with an interactive tutorial is available at sandbox.karrlab.org <a href="http://sandbox.karrlab.org/tree/bpforms"><i class="fi-link"></i></a>.
        </div>
        <div class="large-6 cell">
          <h3>Documentation for the Python API</h3>
          Detailed documentation for the Python API is available at docs.karrlab.org <a href="https://docs.karrlab.org/bpforms"><i class="fi-link"></i></a>.
        </div>
        <div class="large-6 cell">
          <h3>Questions</h3>
          Please contact the Karr Lab <a href="mailto:karr@mssm.edu"><i class="fi-mail"></i></a> with any questions.
        </div>
      </div>

      <div class="grid-x grid-padding-x orange">
        <div class="large-12 cell">
          <h2>Contributing to <i>BpForms</i></h2>
        </div>
        <div class="large-6 cell">
          <h3>Contributing to the alphabets</h3>
          The alphabets are defined by YAML files (see GitHub <a href="https://github.com/KarrLab/bpforms/tree/master/bpforms/alphabet"><i class="fi-link"></i></a>). To contribute to the alphabets or to contribute a new alphabet, please edit or create a YAML file and submit your changes via a Git pull request.
        </div>
        <div class="large-6 cell">
          <h3>Contributing to the software</h3>
          To contribute to the software, please submit a Git pull request.
        </div>
      </div>

      <div class="grid-x grid-padding-x blue">
        <div class="large-12 cell">
          <h2>About <i>BpForms</i></h2>
        </div>
        <div class="large-4 cell">
          <h3>Source code</h3>
          <i>BpForms</i> is available open-source from GitHub <a href="https://github.com/KarrLab/bpforms"><i class="fi-social-github"></i></a>.
        </div>
        <div class="large-4 cell">
          <h3>License</h3>
          <i>BpForms</i> is released under the MIT license <a href="https://github.com/KarrLab/bpforms/blob/master/LICENSE"><i class="fi-link"></i></a>.
        </div>
        <div class="large-4 cell">
          <h3>Citing <i>BpForms</i></h3>
          Lang PF, Chebaro Y &amp; Jonathan R. Karr. BpForms: a toolkit for concretely describing modified DNA, RNA and proteins. arXiv:1903.10042 <a href="https://arxiv.org/abs/1903.10042"><i class="fi-link"></i></a>
        </div>
        <div class="large-4 cell">
          <h3>Team</h3>
          <i>BpForms</i> was developed by Yassmine Chebaro <a href="https://www.linkedin.com/in/yassmine-chebaro-6bb8a05/"><i class="fi-social-linkedin"></i></a>, Jonathan Karr <a href="https://www.karrlab.org/"><i class="fi-link"></i></a> <a href="https://www.linkedin.com/in/jonrkarr/"><i class="fi-social-linkedin"></i></a>, Paul Lang <a href="http://www.dtc.ox.ac.uk/people/17/langp/"><i class="fi-link"></i></a>, and John Sekar <a href="https://www.linkedin.com/in/john-sekar"><i class="fi-social-linkedin"></i></a> in the Karr Lab <a href="https://www.karrlab.org"><i class="fi-link"></i></a> at the Icahn School of Medicine at Mount Sinai <a href="https://www.mssm.edu/"><i class="fi-link"></i></a> in New York, USA.
        </div>
        <div class="large-4 cell">
          <h3>Acknowledgements</h3>
          <i>BpForms</i> was supported by a National Institute of Health P41 award <a href="https://projectreporter.nih.gov/project_info_details.cfm?aid=9460251"><i class="fi-link"></i></a>, a National Institute of Health MIRA R35 award <a href="https://projectreporter.nih.gov/project_info_details.cfm?aid=9509480"><i class="fi-link"></i></a>, and a National Science Foundation INSPIRE award <a href="https://www.nsf.gov/awardsearch/showAward?AWD_ID=1649014"><i class="fi-link"></i></a>.
        </div>        
        <div class="large-4 cell">
          <h3>Questions/comments</h3>
          Please contact Jonathan Karr <a href="mailto:karr@mssm.edu"><i class="fi-mail"></i></a> with any questions or comments.
        </div>
      </div>

    <script src="js/vendor/jquery.js"></script>
    <script src="js/vendor/what-input.js"></script>
    <script src="js/vendor/foundation.js"></script>
    <script src="js/app.js"></script>
  </body>
</html>
