<!doctype html>
<html class="no-js" lang="en" dir="ltr">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>BpForms: toolkit for concretely describing non-canonical DNA, RNA, and proteins</title>
    <link rel="stylesheet" href="css/foundation-icons.css">
    <link rel="stylesheet" href="css/foundation.css">
    <link rel="stylesheet" href="css/app.css">
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway">
  </head>
  <body>
    <div class="grid-container">
      <div class="grid-x grid-padding-x">
        <div class="large-12 cell">
          <h1><i>BpForms</i>: toolkit for concretely describing non-canonical DNA, RNA, and proteins</h1>
        </div>
      </div>

      <div class="grid-x grid-padding-x blue">
        <div class="large-12 cell">
          <p><i>BpForms</i> is a toolkit for unambiguously representing the primary sequence of forms of biopolymers. By concretely representing the primary sequence of biopolymers, <i>BpForms</i> aims to facilitate concrete discussion about DNA modification, post-transcriptional processing, and post-translational processing; facilitate the determination of the structures of biopolymer forms; facilitate the integration of data about DNA modification, post-transcriptional processing, and post-translational processing; and enable whole-cell models <a href="https://www.wholecell.org"><i class="fi-link"></i></a> that represent DNA modification, post-transcriptional processing, and post-translational processing and the functions of modified DNA, RNA, and proteins.</p>
          <p><i>BpForms</i> includes a notation for describing biopolymer forms, as well as this website, a JSON REST API, a command line interface, and a Python API for calculating properties of biopolymer forms. These tools are available open-source under the MIT license.</p>
        </div>
     </div>

      <div class="grid-x grid-padding-x blue">
        <div class="large-12 cell">
            <h2><i>BpForms</i> calculator</h2>
        </div>
        <div class="large-6 cell">
            <a name="form"></a>
            <h3>Enter a biopolymer form</h3>
            <form>
            <label>Select alphabet
                <select id="alphabet">
                </select>
            </label>

            <label>Enter sequence of monomeric forms
                <textarea
                    id="monomer_seq_in"
                    placeholder='AC{m2A}CC[id: dI | structure: "O=C1NC=NC2=C1N=CN2"]G'></textarea>
            </label>

            <label>Select circularity
                <select id="circular">
                  <option value=1>Yes</option>
                  <option value=0 selected>No</option>
                </select>
            </label>

            <label>Optionally, enter a pH to correct the protonation
                <input type="number" id="ph" placeholder="7.4"/>
            </label>

            <label>Optionally, calculate the major tautomer when correcting the protonation
                <select id="major_tautomer">
                  <option value=1>Yes</option>
                  <option value=0 selected>No</option>
                </select>
            </label>


            <div id="warnings" style="color: #da663b;">
            </div>

            <div id="errors" style="color: #da3b60;">
            </div>

            <input id="submit" type="button" class="button" value="Calculate"/>
            </form>
        </div>
        <div class="large-6 cell">
            <h3>Computed properties of the biopolymer form</h3>
            <form class="out">
                <div id="structure-img" style="text-align: center;"></div>
                <label>Length
                    <input type="number" id="length"/>
                </label>
                <label>Structure (SMILES)
                    <textarea id="structure"></textarea>
                </label>
                <label>Chemical formula
                    <input type="text" id="formula"/>
                </label>
                <label>Molecular weight
                    <input type="number" id="mol_wt"/>
                </label>
                <label>Charge
                    <input type="number" id="charge"/>
                </label>
            </form>
        </div>
      </div>

      <div class="grid-x grid-padding-x orange">
        <a name="notation"></a>
        <div class="large-12 cell">
            <h2><i>BpForms</i> notation</h2>
        </div>

        <div class="large-5 cell notation">
          <h3>Overview</h3>
          <p>The <i>BpForms</i> notation represents biopolymers as FASTA sequences augmented with (a) multiple-letter alphabet-defined monomeric forms delimited by curly brackets, (b) user-defined monomeric forms described in square brackets by one or more attributes separated by "|", and (b) additional attributes that can be used to represent circularity and crosslinks.</p>
          <p>The <tt class="attribute">structure</tt>, <tt class="attribute">backbone-bond-atom</tt>, <tt class="attribute">backbone-displaced-atom</tt>, <tt class="attribute">left-bond-atom</tt>, <tt class="attribute">left-displaced-atom</tt>, <tt class="attribute">right-bond-atom</tt>, and <tt class="attribute">right-displaced-atom</tt> monomeric form attributes are required to calculate the chemical formula, molecular weight, and charge. All other attributes are optional.</p>
        </div>
        <div class="large-7 cell example">
          <h3>Examples</h3>
          <div class="grid-x grid-padding-x">
            <div class="large-4 cell">
              <h4>DNA</h4>
              <p style="text-align: center;"><tt>{dI}ACGC</tt></p>
              <img src="img/example/form-DNA.svg"
                style="margin-bottom: 0.5rem;"/>
              <p class="form-properties">
                Length: 5<br/>
                Formula: C<sub>48</sub>H<sub>55</sub>N<sub>20</sub>O<sub>30</sub>P<sub>5</sub><br/>
                Molecular weight: 1546.9<br/>
                Charge: -6
              </p>
              <p>Deoxyinosine at the first position</p>
            </div>
            <div class="large-4 cell">
              <h4>RNA</h4>
              <p style="text-align: center;"><tt>AC{9A}GC</tt></p>
              <img src="img/example/form-RNA.svg"
                style="margin-bottom: 0.5rem;"/>
              <p class="form-properties">
                Length: 5<br/>
                Formula: C<sub>48</sub>H<sub>55</sub>N<sub>20</sub>O<sub>35</sub>P<sub>5</sub><br/>
                Molecular weight: 1626.9<br/>
                Charge: -6
              </p>
              <p>Inosine at the third position</p>
            </div>
            <div class="large-4 cell">
              <h4>Protein</h4>
              <p style="text-align: center;"><tt>ACU[id: "U"]C</tt></p>
              <img src="img/example/form-Protein.svg"
                style="margin-bottom: 0.5rem;"/>
              <p class="form-properties">
                Length: 4<br/>
                Formula: C<sub>12</sub>H<sub>22</sub>N<sub>4</sub>O<sub>5</sub>S<sub>2</sub>Se<sub>1</sub><br/>
                Molecular weight: 445.4<br/>
                Charge: 0
              </p>
              <p>L-selenocysteine at the third position</p>
            </div>
          </div>
        </div>

        <div class="large-5 cell notation">
          <a name="alphabets"></a>
          <h3>Alphabets</h3>
          <p style="margin-bottom: 0px;"><i>BpForms</i> has several pre-built alphabets.</p>
          <ul id="alphabet_descriptions">
          </ul>
        </div>
        <div class="large-7 cell example">
          <h3>Examples</h3>
          <div class="grid-x grid-padding-x">
            <div class="large-4 cell">
              <h4>DNA</h4>
              <p style="text-align: center;"><tt>C{m2A}G{m2C}</tt></p>
              <img src="img/example/alphabet-DNA.svg"
                style="margin-bottom: 0.5rem;"/>
              <p class="form-properties">
                Length: 4<br/>
                Formula: C<sub>39</sub>H<sub>50</sub>N<sub>17</sub>O<sub>24</sub>P<sub>4</sub><br/>
                Molecular weight: 1264.8<br/>
                Charge: -5
              </p>
              <p>2-methyladenine at the second position and 2-methylcytosine at the fourth position</p>
            </div>
            <div class="large-4 cell">
              <h4>RNA</h4>
              <p style="text-align: center;"><tt>A{21C}GC</tt></p>
              <img src="img/example/alphabet-RNA.svg"
                style="margin-bottom: 0.5rem;"/>
              <p class="form-properties">
                Length: 4<br/>
                Formula: C<sub>44</sub>H<sub>60</sub>N<sub>18</sub>O<sub>29</sub>P<sub>4</sub><br/>
                Molecular weight: 1429.0<br/>
                Charge: -4
              </p>
              <p>2-lysidine at the second position</p>
            </div>
            <div class="large-4 cell">
              <h4>Protein</h4>
              <p style="text-align: center;"><tt>{AA0037}E{AA0038}</tt></p>
              <img src="img/example/alphabet-Protein.svg"
                style="margin-bottom: 0.5rem;"/>
              <p class="form-properties">
                Length: 3<br/>
                Formula: C<sub>12</sub>H<sub>18</sub>N<sub>3</sub>O<sub>14</sub>P<sub>2</sub><br/>
                Molecular weight: 490.2<br/>
                Charge: -5
              </p>
              <p>O-phospho-L-serine at the first position and O-phospho-L-threonine at the third position</p>
            </div>
          </div>
        </div>

        <div class="large-5 cell notation">
          <h3>Structures of monomeric forms</h3>
          <p>The <tt class="attribute">structure</tt> monomeric form attribute describes the chemical structure of the inline monomeric form. This attribute is a SMILES-encoded string. Each monomeric form can only have one structure. This attribute is required to calculate the structure of the BpForm.</p>
        </div>
        <div class="large-7 cell example">
          <h3>Example</h3>
          <div class="grid-x grid-padding-x">
            <div class="large-12 cell">
              <p>The text below illustrates how describe the modified DNA nucleobase hypoxanthine, and the image below illustrates the molecule that the text specifies. The atom labels indicate the numbers of the atoms within the molecule. These numbers can be generated with OpenBabel <a href="http://openbabel.org/wiki/Category:Installation"><i class="fi-link"></i></a>.</p>
            </div>
            <div class="large-8 cell syntax">
              <pre>[id: "dI"
    | name: "hypoxanthine"
    | structure: "O=C1NC=NC2=C1N=CN2"
    ]</pre>
            </div>
            <div class="large-4 cell image">
              <img src="img/example/structure-dI.svg"/>
            </div>
          </div>
        </div>

        <div class="large-5 cell notation">
          <h3>Bonds between monomeric forms and backbones</h3>
          <p>The <tt class="attribute">backbone-bond-atom</tt> and <tt class="attribute">backbone-displaced-atom</tt> monomeric form attributes describe the bonds between monomeric forms and their backbone. Each monomeric form can form multiple bonds with the backbone and multiple atoms can be displaced. These attributes are required to calculate the structure of the BpForm.</p>
        </div>
        <div class="large-7 cell example">
          <h3>Example</h3>
          <div class="grid-x grid-padding-x">
            <div class="large-12 cell">
              <p>The example below illustrates how to describe the modified DNA nucleotide deoxyinosine monophosphate. The red atoms in the image indicate the sugar-phosphate backbone.</p>
            </div>
            <div class="large-8 cell syntax">
              <pre>[id: "dI"
    | name: "hypoxanthine"
    | structure: "O=C1NC=NC2=C1N=CN2"
    | backbone-bond-atom: N10
    | backbone-displaced-atom: H10
    ]</pre>
            </div>
            <div class="large-4 cell image">
              <img src="img/example/monomer-backbone-bonds-dI.svg"/>
            </div>
          </div>
        </div>

        <div class="large-5 cell notation">
          <h3>Bonds between adjacent monomeric forms</h3>
          <p>The <tt class="attribute">left-bond-atom</tt>, <tt class="attribute">left-displaced-atom</tt>, <tt class="attribute">right-bond-atom</tt>, and <tt class="attribute">right-displaced-atom</tt> monomeric form attributes describe the bonds between successive monomeric forms. Each monomeric form can have multiple bonds and multiple displaced atoms. These attributes are required to calculate the structure of the BpForm.</p>
        </div>
        <div class="large-7 cell example">
          <h3>Example</h3>
          <div class="grid-x grid-padding-x">
            <div class="large-12 cell">
              <p>The example below illustrates how to describe the modified protein residue N<sup>5</sup>-methyl-L-arginine. The green atoms indicate the C terminus. The blue atoms indicate N terminus.</p>
            </div>
            <div class="large-8 cell syntax">
              <pre>[id: "AA0305"
    | name: "N5-methyl-L-arginine"
    | structure: "O=C[C@H](CCCN(C(=[NH2])N)C)[NH3+]"
    | backbone-bond-atom: C2
    | backbone-displaced-atom: H2
    | right-bond-atom: C2
    | left-bond-atom: N15-1
    | left-displaced-atom: H15+1
    | left-displaced-atom: H15+1
    ]</pre>
            </div>
            <div class="large-4 cell image">
              <img src="img/example/left-right-bonds-AA0305.svg"/>
            </div>
            <div class="large-12 cell" style="margin-top:1rem;">
              <p>The example below illustrates how to describe a polymer with N<sup>5</sup>-methyl-L-arginine at the third position. The green lines indicate the bonds between the successive monomeric forms. The black labels indicate the positions of monomeric forms within the sequence.</p>
            </div>
            <div class="large-8 cell syntax">
              <pre>A[id: "AA0305"
    | name: "N5-methyl-L-arginine"
    | structure: "O=C[C@H](CCCN(C(=[NH2])N)C)[NH3+]"
    | backbone-bond-atom: C2
    | backbone-displaced-atom: H2
    | right-bond-atom: C2
    | left-bond-atom: N15-1
    | left-displaced-atom: H15+1
    | left-displaced-atom: H15+1
    ]A</pre>
            </div>
            <div class="large-4 cell image">
              <img src="img/example/left-right-bonds-A-AA0305-A.svg"/>
            </div>
          </div>
        </div>

        <div class="large-5 cell notation">
          <h3>Bonds that create circular polymers</h3>
          <p>The <tt class="attribute">circular</tt> polymer attribute indicates that there is a bond between the last and first monomeric forms, creating a circular polymer.</p>
        </div>
        <div class="large-7 cell example">
          <h3>Example</h3>
          <div class="grid-x grid-padding-x">
            <div class="large-12 cell">
              <p>The example below illustrates how to describe the circular DNA dimer of the DNA nucleotides deoxyadenosine monophosphaate and dexocytosine monophosphate. The red atoms indicates the sugar-phosphate backbones. The green lines indicate the bonds between successive sugar-phosphate backbones. The black labels indicate the positions of monomeric forms within the sequence.</p>
            </div>
            <div class="large-8 cell syntax">
              <pre>AC | circular</pre>
            </div>
            <div class="large-4 cell image">
              <img src="img/example/circular-DNA-AC.svg"/>
            </div>
          </div>
        </div>

        <div class="large-5 cell notation">
          <h3>Crosslinks between monomeric forms</h3>
          <p>The <tt class="attribute">crosslink</tt> polymer attribute indicates that there is a bond between two non-adjacent monomeric forms. Polymers can have zero, one, or more crosslinks. This attribute can be used to describe intrastran crosslinks in DNA, disulfide bonds between cysteines in proteins, and other bonds.</p>
        </div>
        <div class="large-7 cell example">
          <h3>Example</h3>
          <div class="grid-x grid-padding-x">
            <div class="large-12 cell">
              <p>The example below illustrates how to describe a tripeptide with a disulfide bond. The blue line indicates the disulfide bond (crosslink). The green lines indicate the bonds between the successive residues. The black labels indicate the positions of monomeric forms within the sequence.</p>
            </div>
            <div class="large-8 cell syntax">
              <pre>CAC | crosslink: [
    left-bond-atom: 1S1 |
    left-displaced-atom: 1H1 |
    right-bond-atom: 1S1 |
    right-displaced-atom: 1H1
]</pre>
            </div>
            <div class="large-4 cell image">
              <img src="img/example/crosslink-protein-sulfide-bond.svg"/>
            </div>
          </div>
        </div>

        <div class="large-5 cell notation">
          <h3>Uncertainty about the primary sequence</h3>
          <p class="before_list"><i>BpForms</i> can represent two types of uncertainty in the primary sequences of biopolymer forms.</p>
          <ul>
          <li>The <tt class="attribute">delta-mass</tt> <tt class="attribute">delta-charge</tt> monomeric form attributes describe uncertainty in the chemical identity of the monomeric form.</li>
          <li>The <tt class="attribute">position</tt> monomeric form attribute describes uncertainty in the position of the monomeric form within the sequence.</i>
          </ul>
        </div>
        <div class="large-7 cell example">
          <h3>Examples</h3>
          <ul>
          <li><tt>[id: "dAMP" | delta-mass: 1 | delta-charge: 1]</tt>: indicates the presence of an additional proton whose exact location is not known.</li>
          <li><tt>[id: "dI" | position: 2-3]</tt>: indicates that deoxyinosine may occur anywhere between the second and third position.</li>
          </ul>
        </div>


        <div class="large-5 cell notation">
          <h3>Metadata about monomeric forms</h3>
          <p class="before_list"><i>BpForms</i> can represent several types of metadata about monomeric forms.</p>
          <ul>
            <li>The <tt class="attribute">id</tt> and <tt class="attribute">name</tt> monomeric form attributes are human-readable labels for monomeric forms. Only one id and one name is allowed per monomeric form.</li>
            <li>The <tt class="attribute">synonym</tt> monomeric form attribute is an additional human-readable label. Monomeric forms can have multiple synonyms.</li>
            <li>The <tt class="attribute">identifier</tt> monomeric form attribute indicates entries in databases and ontologies which are equivalent to the monomeric form. Monomeric forms can have multiple identifiers. The namespace and id of each identifer must be separated by a "/".</li>
            <li>The <tt class="attribute">comments</tt> monomeric form attribute describes additional information about the monomeric form. Monomeric forms can only have one comment.</li>
          </ul>
        </div>
        <div class="large-7 cell example">
          <h3>Examples</h3>
          <ul>
          <li><tt>[id: "dI" | name: "deoxyinosine"]</tt>: represents the id and name of deoxyinosine.</li>
          <li><tt>[id: "dI" | synonym: "deoxyinosine" | synonym: "2'-deoxyinosine"]</tt>: represents multiple synonyms of deoxyinosine.</li>
          <li><tt>[id: "dI" | identifier: "CHEBI:28997" @ "chebi" | identifier: "65058" @ "pubchem.compound"]</tt>: represents equivalent entries in ChEBI and PubChem to deoxyinosine.</li>
          <li><tt>[id: "dI" | comments: "A purine 2'-deoxyribonucleoside that is inosine ..."]</tt>: represents comments about deoxyinosine.</li>
          </ul>
        </div>
      </div>

      <div class="grid-x grid-padding-x blue">
        <div class="large-12 cell">
          <h2>Resources for reconstructing biopolymer forms</h2>
        </div>
        <div class="large-6 cell">
          <h3>DNA resources</h3>
          <ul>
            <li><i>DNAMod</i> <a href="https://dnamod.hoffmanlab.org/"><i class="fi-link"></i></a>: Database of non-canonical DNA nucleobases</li>
            <li><i>MethDB</i> <a href="http://www.methdb.net/"><i class="fi-link"></i></a>: Database of non-canonical DNA</li>
            <li><i>MethSMRT</i> <a href="http://sysbio.gzzoc.com/methsmrt/"><i class="fi-link"></i></a>: Database of non-canonical DNA</li>
          </ul>

          <h3>RNA resources</h3>
          <ul>
            <li><i>MODOMICS</i> <a href="http://modomics.genesilico.pl/modifications/"><i class="fi-link"></i></a>: Database of non-canonical RNA nucleosides</li>
            <li><i>RMBase</i> <a href="http://rna.sysu.edu.cn/rmbase/"><i class="fi-link"></i></a>: Database of modified RNA</li>
            <li><i>RNA Modification Database</i> <a href="https://mods.rna.albany.edu/"><i class="fi-link"></i></a>: Database of modified RNA</li>
          </ul>

          <h3>Drawing structures of monomeric forms</h3>
          <ul>
            <li><i>ChemAxon Marvin</i> <a href="https://chemaxon.com/products/marvin"><i class="fi-link"></i></a>: Software for drawing structures of monomeric forms</li>
            <li><i>OpenBabel</i> <a href="http://openbabel.org/wiki/Category:Installation"><i class="fi-link"></i></a>: Software for calculating the numbers of the atoms in monomeric forms</li>
          </ul>
        </div>
        <div class="large-6 cell">
          <h3>Protein resources</h3>
          <ul>
            <li><i>dbPTM</i> <a href="http://dbptm.mbc.nctu.edu.tw/"><i class="fi-link"></i></a>: Database of non-canonical amino acids</li>
            <li><i>Delta Mass</i> <a href="https://abrf.org/delta-mass"><i class="fi-link"></i></a>: Database of modified amino acids</li>
            <li><i>FindMod</i> <a href="https://web.expasy.org/findmod/findmod_masses.html"><i class="fi-link"></i></a>: Database of post-translational modifications</li>
            <li><i>ProForma</i> <a href="https://www.topdownproteomics.org/resources/proforma/"><i class="fi-link"></i></a>: Notation for protein forms.  Note, this notation is not unambiguous. This limits its abiltiy to facilitate data integration and the calculation of properties of protein forms.</li>
            <li><i>PDB Chemical Components</i> <a href="http://www.wwpdb.org/data/ccd"><i class="fi-link"></i></a>: Database of modified amino acids</li>
            <li><i>PDB in Europe Chemical Components</i> <a href="https://www.ebi.ac.uk/pdbe/pdb-component-library/index.html>"><i class="fi-link"></i></a>: Database of modified amino acids</li>
            <li><i>Protein Ontology</i> <a href="https://pir.georgetown.edu/pro/"><i class="fi-link"></i></a>: Database of modified proteins</li>
            <li><i>PSIMOD</i> <a href="https://bioportal.bioontology.org/ontologies/PSIMOD"><i class="fi-link"></i></a>: Ontology of non-canonical amino acids</li>
            <li><i>RESID</i> <a href="https://pir.georgetown.edu/resid/"><i class="fi-link"></i></a>: Database of non-canonical protein residues</li>
            <li><i>UniMod</i> <a href="http://www.unimod.org/"><i class="fi-link"></i></a>: Database of non-canonical amino acids</li>
            <li><i>UniProt</i> <a href="https://www.uniprot.org/"><i class="fi-link"></i></a>: Database of modified amino acids in proteins</li>
            <li><i>UniProt Controlled Vocabulary of Posttranslational Modifications</i> <a href="https://www.uniprot.org/docs/ptmlist"><i class="fi-link"></i></a>: Database of modified amino acids</li>
          </ul>
        </div>
      </div>

      <div class="grid-x grid-padding-x orange">
        <div class="large-12 cell">
          <h2>Integrating BpForms into pathway databases, models, and <i>in silico</i> designs</h2>
        </div>
        <div class="large-4 cell">
          <h3>Knowledge (BioPAX <a href="https://www.biopax.org/"><i class="fi-link"></i></a>)</h3>
          <p><i>BpForms</i> can be embedded into <tt>sequence</tt> children of instances of <tt>DNAReference</tt>, <tt>RNAReference</tt>, and <tt>ProteinReference</tt>.
          </p>
        </div>
        <div class="large-8 cell example">
          <h3>Examples</h3>
          <h4>DNA: <span style="font-weight: normal;"><i>E. coli</i> K-12 MG1655 Dam 6-methyladenine sites (701..914) involved in host recognition <a href="https://www.ncbi.nlm.nih.gov/nuccore/NC_000913.3"><i class="fi-link"></i></a></span></h4>
          <pre class="small">...
  &lt;bp:DNA&gt;
    &lt;bp:entityReference&gt;
      &lt;bp:DNAReference&gt;
        &lt;bp:sequence 
          rdf:datatype="http://www.w3.org/2001/XMLSchema#string" 
          rdf:about="http://edamontology.org/format_3909#dna"&gt;
          <span class="bpform">...
          TGATTTGCCGTGGCGAGAAAATGTCG<span class="modification">{a}</span>TCGCCATTATGGCCGGCGTATTAGAAGCGCGCGGTCACAAC
          GTTACTGTTATCG<span class="modification">{a}</span>TCCGGTCGAAAAACTGCTGGCAGTGGGGCATTACCTCGAATCTACCGTCGATAT
          TGCTGAGTCCACCCGCCGTATTGCGGCAAGCCGCATTCCGGCTG<span class="modification">{a}</span>TCACATGGTGCTGATGGCAGGTT
          ...</span>
        &lt;/bp:sequence&gt;
      &lt;/bp:DNAReference&gt;
    &lt;/bp:entityReference&gt;
  &lt;/bp:DNA&gt;
...
</pre>

          <h4>RNA: <span style="font-weight: normal;">Modifications of <i>B. subtilis</i> tRNA<sup>UGC</sup> involved in stability <a href="http://modomics.genesilico.pl/sequences/list/?type_field=tRNA&subtype=Ala&species=4&display=Display&nomenclature=abbrev"><i class="fi-link"></i></a></span></h4>
          <pre class="small">...
  &lt;bp:RNA&gt;
    &lt;bp:entityReference&gt;
      &lt;bp:RNAReference&gt;
        &lt;bp:sequence
          rdf:datatype="http://www.w3.org/2001/XMLSchema#string"
          rdf:about="http://edamontology.org/format_3909#rna"&gt;
          <span class="bpform">GGAGCCUUAGCUCAGC<span class="modification">{8U}</span>GGGAGAGCGCCUGCUU<span class="modification">{501U}</span>GC<span class="modification">{6A}</span>CGCAGGAG<span class="modification">{7G}</span>UCAGCGG<span class="modification">{5U}</span><span class="modification">{9U}</span>CGAUCCCGCUAGGCUCCA
          CCA</span>
        &lt;/bp:sequence&gt;
      &lt;/bp:RNAReference&gt;
    &lt;/bp:entityReference&gt;
  &lt;/bp:RNA&gt;
...</pre>

          <h4>Protein: <span style="font-weight: normal;">Modifications of <i>H. sapiens</i> MAPK3 involved in signaling <a href="https://www.uniprot.org/uniprot/P27361#ptm_processing"><i class="fi-link"></i></a></span></h4>
          <pre class="small">...
  &lt;bp:Protein&gt;
    &lt;bp:entityReference&gt;
      &lt;bp:ProteinReference&gt;
        &lt;bp:sequence
          rdf:datatype="http://www.w3.org/2001/XMLSchema#string"
          rdf:about="http://edamontology.org/format_3909#protein"&gt;
          <span class="bpform">M<span class="modification">{AA0041}</span>AAAAQGGGGGEPRRTEGVGPGVPGEVEMVKGQPFDVGPRYTQLQYIGEGAYGMVSSAYDHVRKTRVAIKKISPFEHQTYCQRTL
          REIQILLRFRHENVIGIRDILRASTLEAMRDVYIVQDLMETDLYKLLKSQQLSNDHICYFLYQILRGLKYIHSANVLHRDLKPSNLLINTTCD
          LKICDFGLARIADPEHDH<span class="modification">{AA0038}</span>GFL<span class="modification">{AA0038}</span>E<span class="modification">{AA0039}</span>VA<span class="modification">{AA0038}</span>RWYRAPEIMLNSKGYTKSIDIWSVGCILAEMLSNRPI
          FPGKHYLDQLNHILGILGSPSQEDLNCIINMKARNYLQSLPSKTKVAWAKLFPKSDSKALDLLDRMLTFNPNKRITVEEALAHPYLEQYYDPT
          DEPVAEEPFTFAMELDDLPKERLKELIFQETARFQPGVLEAP</span>
        &lt;/bp:sequence&gt;
      &lt;/bp:ProteinReference&gt;
    &lt;/bp:entityReference&gt;
  &lt;/bp:Protein&gt;
...</pre>

        </div>

        <div class="large-4 cell">
          <h3>Models (SBML <a href="https://sbml.org"><i class="fi-link"></i></a>)</h3>
          <p>The <tt class="attribute">annotation</tt> attribute of <tt class="attribute">species</tt> can be used to describe species with BpForms.</p>
        </div>
        <div class="large-8 cell example">
          <h3>Example</h3>
<pre class="small">...
  &lt;species&gt;
    &lt;annotation&gt;
      &lt;bpforms:bpform xmlns:bpforms="https://www.bpforms.org/ns"&gt;
        &lt;bpforms:alphabet&gt;dna&lt;/bpforms:alphabet&gt;
        &lt;bpforms:sequence&gt;A{m2C}GT&lt;/bpforms:sequence&gt;
      &lt;/bpforms:bpform&gt;
    &lt;/annotation&gt;
  &lt;/species&gt;
...</pre>
        </div>

        <div class="large-4 cell">
          <h3><i>In silico</i> designs (SBOL <a href="http://sbolstandard.org/"><i class="fi-link"></i></a>)</h3>
          <p>This encoding URI can be used to describe components in <i>in silico</i> designs: <a href="http://edamontology.org/format_3909">http://edamontology.org/format_3909</a></p>
        </div>
        <div class="large-8 cell example">
          <h3>Example</h3>
<pre class="small">...
  &lt;sbol:Sequence&gt;
    &lt;sbol:encoding rdf:resource="http://edamontology.org/format_3909"/&gt;
    &lt;sbol:elements&gt;A{m2C}GT&lt;/sbol:elements&gt;
  &lt;/sbol:Sequence&gt;
...</pre>
        </div>
      </div>

      <div class="grid-x grid-padding-x blue">
        <div class="large-12 cell">
          <h2><i>BpForm</i> interfaces</h2>
        </div>
        <div class="large-4 cell">
          <h3>Webform</h3>
          The webform above <a href="#form"><i class="fi-link"></i></a> can be used to validate BpForms and calculate their properties.
        </div>
        <div class="large-4 cell">
          <h3>JSON REST API</h3>
          The <i>BpForms</i> JSON REST API is available at <a href="https://bpforms.org/api">https://bpforms.org/api</a>.
        </div>
        <div class="large-4 cell">
          <h3>Command line interface</h3>
          The <i>BpForms</i> command line interface is available from PyPI <a href="https://pypi.python.org/pypi/bpforms"><i class="fi-link"></i></a>.
        </div>
        <div class="large-4 cell">
          <h3>Python package</h3>
          The <i>BpForms</i> Python package is available from PyPI <a href="https://pypi.python.org/pypi/bpforms"><i class="fi-link"></i></a>.
        </div>
        <div class="large-4 cell">
          <h3>Source code</h3>
          <i>BpForms</i> is available open-source from GitHub <a href="https://github.com/KarrLab/bpforms"><i class="fi-social-github"></i></a>.
        </div>
      </div>

      <div class="grid-x grid-padding-x orange">
        <div class="large-12 cell">
          <h2>Tutorials, documentation, and help</h2>
        </div>
        <div class="large-6 cell">
          <h3>Documentation for the notation</h3>
          Documentation for the notation is available above <a href="#notation"><i class="fi-link"></i></a>. The grammar for the notation is available at GitHub <a href="https://raw.githubusercontent.com/KarrLab/bpforms/master/bpforms/grammar.lark"><i class="fi-link"></i></a>
        </div>
        <div class="large-6 cell">
          <h3>Definitions of the monomeric forms in the alphabets</h3>
          Documentation for the alphabets is available above <a href="#alphabets"><i class="fi-link"></i></a>. This includes images and detailed information about each monomeric form in each alphabet.
        </div>
        <div class="large-6 cell">
          <h3>Query builder for the REST API</h3>
          A visual interface for building REST queries is available at bpforms.org/api <a href="https://www.bpforms.org/api"><i class="fi-link"></i></a>.
        </div>
        <div class="large-6 cell">
          <h3>Documentation for the REST API</h3>
          Documentation for the REST API is available at bpforms.org/api <a href="https://www.bpforms.org/api"><i class="fi-link"></i></a>.
        </div>
        <div class="large-6 cell">
          <h3>Documentation for the command line program</h3>
          Documentation for the command line program is available inline by running <tt>bpforms --help</tt>.
        </div>
        <div class="large-6 cell">
          <h3>Tutorial for the Python API</h3>
          A Jupyter notebook with an interactive tutorial is available at sandbox.karrlab.org <a href="http://sandbox.karrlab.org/tree/bpforms"><i class="fi-link"></i></a>.
        </div>
        <div class="large-6 cell">
          <h3>Documentation for the Python API</h3>
          Detailed documentation for the Python API is available at docs.karrlab.org <a href="https://docs.karrlab.org/bpforms"><i class="fi-link"></i></a>.
        </div>
        <div class="large-6 cell">
          <h3>Questions</h3>
          Please contact the Karr Lab <a href="mailto:karr@mssm.edu"><i class="fi-mail"></i></a> with any questions.
        </div>
      </div>

      <div class="grid-x grid-padding-x blue">
        <div class="large-12 cell">
          <h2>Contributing to <i>BpForms</i></h2>
        </div>
        <div class="large-6 cell">
          <h3>Contributing to the alphabets</h3>
          The alphabets are defined by YAML files (see GitHub <a href="https://github.com/KarrLab/bpforms/tree/master/bpforms/alphabet"><i class="fi-link"></i></a>). To contribute to the alphabets or to contribute a new alphabet, please edit or create a YAML file and submit your changes via a Git pull request.
        </div>
        <div class="large-6 cell">
          <h3>Contributing to the software</h3>
          To contribute to the software, please submit a Git pull request.
        </div>
      </div>

      <div class="grid-x grid-padding-x orange">
        <div class="large-12 cell">
          <h2>About <i>BpForms</i></h2>
        </div>
        <div class="large-4 cell">
          <h3>Citing <i>BpForms</i></h3>
          Lang PF, Chebaro Y &amp; Jonathan R. Karr. BpForms: a toolkit for concretely describing modified DNA, RNA and proteins. arXiv:1903.10042 <a href="https://arxiv.org/abs/1903.10042"><i class="fi-link"></i></a>
        </div>
        <div class="large-4 cell">
          <h3>Team</h3>
          <i>BpForms</i> was developed by Jonathan Karr <a href="https://www.karrlab.org/"><i class="fi-link"></i></a> <a href="https://www.linkedin.com/in/jonrkarr/"><i class="fi-social-linkedin"></i></a>, Yassmine Chebaro <a href="https://www.linkedin.com/in/yassmine-chebaro-6bb8a05/"><i class="fi-social-linkedin"></i></a>, and Paul Lang <a href="http://www.dtc.ox.ac.uk/people/17/langp/"><i class="fi-link"></i></a> in the Karr Lab <a href="https://www.karrlab.org"><i class="fi-link"></i></a> at the Icahn School of Medicine at Mount Sinai <a href="https://www.mssm.edu/"><i class="fi-link"></i></a> in New York, USA.
        </div>
        <div class="large-4 cell">
          <h3>Acknowledgements</h3>
          <i>BpForms</i> was supported by a National Institute of Health P41 award <a href="https://projectreporter.nih.gov/project_info_details.cfm?aid=9460251"><i class="fi-link"></i></a>, a National Institute of Health MIRA R35 award <a href="https://projectreporter.nih.gov/project_info_details.cfm?aid=9509480"><i class="fi-link"></i></a>, and a National Science Foundation INSPIRE award <a href="https://www.nsf.gov/awardsearch/showAward?AWD_ID=1649014"><i class="fi-link"></i></a>.
        </div>
        <div class="large-4 cell">
          <h3>License</h3>
          <i>BpForms</i> is released under the MIT license <a href="https://github.com/KarrLab/bpforms/blob/master/LICENSE"><i class="fi-link"></i></a>.
        </div>
        <div class="large-4 cell">
          <h3>Questions/comments</h3>
          Please contact Jonathan Karr <a href="mailto:karr@mssm.edu"><i class="fi-mail"></i></a> with any questions or comments.
        </div>
      </div>

    <script src="js/vendor/jquery.js"></script>
    <script src="js/vendor/what-input.js"></script>
    <script src="js/vendor/foundation.js"></script>
    <script src="js/app.js"></script>
  </body>
</html>
